Grid/qcd/utils/GaugeGroup.h

```diff
@@ -0,0 +1,512 @@
+    for (int i = 2; i <= 12; ++i) {
+      nfac = nfac / RealD(i);  // 1/2, 1/2.3 ...
+      xn = xn * x;             // x2, x3,x4....
+      ex = ex + xn * nfac;     // x2/2!, x3/3!....
+    }
+  }
+    
+  template <int N>                  // Projects on the general groups U(N), Sp(2N)xZ2 i.e. determinant il allowed a complex phase.
+  static void ProjectOnGeneralGroup(Lattice<iVector<iScalar<iMatrix<vComplexD, N> >, Nd> > &U) {
+    for (int mu = 0; mu < Nd; mu++) {
+      auto Umu = PeekIndex<LorentzIndex>(U, mu);
+      ProjectOnGeneralGroup(Umu);
+    }
+  }
+       
+  template <int N>
+  static void ProjectOnGeneralGroup(Lattice<iScalar<iScalar<iMatrix<vComplexD, N> > > > &Umu) {
+      ProjectOnGeneralGroup(Umu, group_name());
+  }
+    
+  template <class vtype>
+  static void ProjectOnGeneralGroup(iScalar<vtype> &r) {
+      ProjectOnGeneralGroup(r, group_name());
+  }
+
+  template <class vtype, int N>
+  static void ProjectOnGeneralGroup(iVector<vtype,N> &r) {
+      ProjectOnGeneralGroup(r, group_name());
+  }
+
+  template <class vtype,int N, typename std::enable_if< GridTypeMapper<vtype>::TensorLevel == 0 >::type * =nullptr>
+  static void ProjectOnGeneralGroup(iMatrix<vtype,N> &arg) {
+      ProjectOnGeneralGroup(arg, group_name());
+  }
+
+  template <int N>       // Projects on SU(N), Sp(2N)
+  static void ProjectOnSpecialGroup(Lattice<iScalar<iScalar<iMatrix<vComplexD, N> > > > &Umu) {
+       ProjectOnGeneralGroup(Umu);
+       auto det = Determinant(Umu);
+
+       det = conjugate(det);
+
+       for (int i = 0; i < N; i++) {
+           auto element = PeekIndex<ColourIndex>(Umu, N - 1, i);
+           element = element * det;
+           PokeIndex<ColourIndex>(Umu, element, Nc - 1, i);
+       }
+   }
+
+    template <int N>    // reunitarise, resimplectify... previously ProjectSUn
+    static void ProjectOnSpecialGroup(Lattice<iVector<iScalar<iMatrix<vComplexD, N> >, Nd> > &U) {
+      // Reunitarise
+      for (int mu = 0; mu < Nd; mu++) {
+        auto Umu = PeekIndex<LorentzIndex>(U, mu);
+        ProjectOnSpecialGroup(Umu);
+        PokeIndex<LorentzIndex>(U, Umu, mu);
+      }
+    }
+
+};
+
+template <int N>
+LatticeComplexD Determinant(
+    const Lattice<iScalar<iScalar<iMatrix<vComplexD, N> > > > &Umu) {
+  GridBase *grid = Umu.Grid();
+  auto lvol = grid->lSites();
+  LatticeComplexD ret(grid);
+
+  autoView(Umu_v, Umu, CpuRead);
+  autoView(ret_v, ret, CpuWrite);
+  thread_for(site, lvol, {
+    Eigen::MatrixXcd EigenU = Eigen::MatrixXcd::Zero(N, N);
+    Coordinate lcoor;
+    grid->LocalIndexToLocalCoor(site, lcoor);
+    iScalar<iScalar<iMatrix<ComplexD, N> > > Us;
+    peekLocalSite(Us, Umu_v, lcoor);
+    for (int i = 0; i < N; i++) {
+      for (int j = 0; j < N; j++) {
+        EigenU(i, j) = Us()()(i, j);
+      }
+    }
+    ComplexD det = EigenU.determinant();
+    pokeLocalSite(det, ret_v, lcoor);
+  });
+  return ret;
+}
+
+template <int N>
+static void ProjectSUn(
```

Comment:
Code duplication. Should roughly be
```suggestion
template <int N>
static void ProjectSUn(Lattice<iScalar<iScalar<iMatrix<vComplexD, N> > > > &Umu) {
    GaugeGroup<N,GroupName::SU>::ProjectOnSpecialGroup(Umu);
}
```
or even better (haven't tried but something like this should work)
```suggestion
template <int N>
using ProjectSUn = GaugeGroup<N,GroupName::SU>::ProjectOnSpecialGroup;
```

Reply:
compiler seems to also want a typename before GaugeGroup

will implement

Reply:
I've implemented but I've been sloppy I still need to pass a default argument to the template otherwise core/Test_reunitarise fails

`error: use of alias template 'ProjectSUn' requires template arguments
  
ProjectSUn(UU);
`

Reply:
now it works I think. I went with your first idea